File: basic_dsa_three.java
/**
 * Basic DSA Problems - Two More Fundamental Questions
 * Simple and essential DSA problems for beginners
 */
public class basic_dsa_three {
    
    // Problem 1: Check if Array is Sorted
    // Check whether an array is sorted in ascending order
    /**
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static boolean isSorted(int[] arr) {
        if (arr.length <= 1) {
            return true;
        }
        
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] < arr[i - 1]) {
                return false;
            }
        }
        
        return true;
    }
    
    // Problem 2: Find Second Largest Element in Array
    // Find the second largest element in an array
    /**
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static Integer findSecondLargest(int[] arr) {
        if (arr.length < 2) {
            return null;
        }
        
        int largest = Integer.MIN_VALUE;
        int secondLargest = Integer.MIN_VALUE;
        
        for (int num : arr) {
            if (num > largest) {
                secondLargest = largest;
                largest = num;
            } else if (num > secondLargest && num != largest) {
                secondLargest = num;
            }
        }
        
        if (secondLargest == Integer.MIN_VALUE) {
            return null; // All elements are same or only one unique element
        }
        
        return secondLargest;
    }
    
    // Helper method to print array
    private static void printArray(int[] arr) {
        System.out.print("[");
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
            if (i < arr.length - 1) {
                System.out.print(", ");
            }
        }
        System.out.print("]");
    }
    
    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Check if Array is Sorted");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - Sorted array
        int[] arr1 = {1, 2, 3, 4, 5};
        boolean result1 = isSorted(arr1);
        System.out.print("Array: ");
        printArray(arr1);
        System.out.println();
        System.out.println("Is sorted: " + result1);
        
        // Test Case 2 - Unsorted array
        int[] arr2 = {5, 2, 8, 1, 9};
        boolean result2 = isSorted(arr2);
        System.out.print("\nArray: ");
        printArray(arr2);
        System.out.println();
        System.out.println("Is sorted: " + result2);
        
        // Test Case 3 - Single element
        int[] arr3 = {10};
        boolean result3 = isSorted(arr3);
        System.out.print("\nArray: ");
        printArray(arr3);
        System.out.println();
        System.out.println("Is sorted: " + result3);
        
        // Test Case 4 - Empty array
        int[] arr4 = {};
        boolean result4 = isSorted(arr4);
        System.out.print("\nArray: ");
        printArray(arr4);
        System.out.println();
        System.out.println("Is sorted: " + result4);
        
        // Test Case 5 - Duplicate elements (sorted)
        int[] arr5 = {1, 2, 2, 3, 4};
        boolean result5 = isSorted(arr5);
        System.out.print("\nArray: ");
        printArray(arr5);
        System.out.println();
        System.out.println("Is sorted: " + result5);
        
        // Test Case 6 - Descending order (not sorted for ascending)
        int[] arr6 = {5, 4, 3, 2, 1};
        boolean result6 = isSorted(arr6);
        System.out.print("\nArray: ");
        printArray(arr6);
        System.out.println();
        System.out.println("Is sorted: " + result6);
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Problem 2: Find Second Largest Element in Array");
        System.out.println("=".repeat(60));
        
        // Test Case 1
        int[] arr7 = {10, 5, 8, 20, 15};
        Integer result7 = findSecondLargest(arr7);
        System.out.print("Array: ");
        printArray(arr7);
        System.out.println();
        System.out.println("Second largest: " + (result7 != null ? result7 : "None"));
        
        // Test Case 2
        int[] arr8 = {1, 2, 3, 4, 5};
        Integer result8 = findSecondLargest(arr8);
        System.out.print("\nArray: ");
        printArray(arr8);
        System.out.println();
        System.out.println("Second largest: " + (result8 != null ? result8 : "None"));
        
        // Test Case 3 - Duplicate largest
        int[] arr9 = {10, 10, 8, 7, 5};
        Integer result9 = findSecondLargest(arr9);
        System.out.print("\nArray: ");
        printArray(arr9);
        System.out.println();
        System.out.println("Second largest: " + (result9 != null ? result9 : "None"));
        
        // Test Case 4 - All same elements
        int[] arr10 = {5, 5, 5, 5};
        Integer result10 = findSecondLargest(arr10);
        System.out.print("\nArray: ");
        printArray(arr10);
        System.out.println();
        System.out.println("Second largest: " + (result10 != null ? result10 : "None"));
        
        // Test Case 5 - Two elements
        int[] arr11 = {10, 5};
        Integer result11 = findSecondLargest(arr11);
        System.out.print("\nArray: ");
        printArray(arr11);
        System.out.println();
        System.out.println("Second largest: " + (result11 != null ? result11 : "None"));
        
        // Test Case 6 - Negative numbers
        int[] arr12 = {-5, -2, -10, -8, -1};
        Integer result12 = findSecondLargest(arr12);
        System.out.print("\nArray: ");
        printArray(arr12);
        System.out.println();
        System.out.println("Second largest: " + (result12 != null ? result12 : "None"));
    }
}

File: basic_dsa_three.py
"""
Basic DSA Problems - Two More Fundamental Questions
Simple and essential DSA problems for beginners
"""


# Problem 1: Check if Array is Sorted
# Check whether an array is sorted in ascending order
def is_sorted(arr):
    """
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if len(arr) <= 1:
        return True
    
    for i in range(1, len(arr)):
        if arr[i] < arr[i - 1]:
            return False
    
    return True


# Problem 2: Find Second Largest Element in Array
# Find the second largest element in an array
def find_second_largest(arr):
    """
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if len(arr) < 2:
        return None
    
    largest = float('-inf')
    second_largest = float('-inf')
    
    for num in arr:
        if num > largest:
            second_largest = largest
            largest = num
        elif num > second_largest and num != largest:
            second_largest = num
    
    if second_largest == float('-inf'):
        return None  # All elements are same or only one unique element
    
    return second_largest


# Test cases
if __name__ == "__main__":
    print("=" * 60)
    print("Problem 1: Check if Array is Sorted")
    print("=" * 60)
    
    # Test Case 1 - Sorted array
    arr1 = [1, 2, 3, 4, 5]
    result1 = is_sorted(arr1)
    print(f"Array: {arr1}")
    print(f"Is sorted: {result1}")
    
    # Test Case 2 - Unsorted array
    arr2 = [5, 2, 8, 1, 9]
    result2 = is_sorted(arr2)
    print(f"\nArray: {arr2}")
    print(f"Is sorted: {result2}")
    
    # Test Case 3 - Single element
    arr3 = [10]
    result3 = is_sorted(arr3)
    print(f"\nArray: {arr3}")
    print(f"Is sorted: {result3}")
    
    # Test Case 4 - Empty array
    arr4 = []
    result4 = is_sorted(arr4)
    print(f"\nArray: {arr4}")
    print(f"Is sorted: {result4}")
    
    # Test Case 5 - Duplicate elements (sorted)
    arr5 = [1, 2, 2, 3, 4]
    result5 = is_sorted(arr5)
    print(f"\nArray: {arr5}")
    print(f"Is sorted: {result5}")
    
    # Test Case 6 - Descending order (not sorted for ascending)
    arr6 = [5, 4, 3, 2, 1]
    result6 = is_sorted(arr6)
    print(f"\nArray: {arr6}")
    print(f"Is sorted: {result6}")
    
    print("\n" + "=" * 60)
    print("Problem 2: Find Second Largest Element in Array")
    print("=" * 60)
    
    # Test Case 1
    arr7 = [10, 5, 8, 20, 15]
    result7 = find_second_largest(arr7)
    print(f"Array: {arr7}")
    print(f"Second largest: {result7}")
    
    # Test Case 2
    arr8 = [1, 2, 3, 4, 5]
    result8 = find_second_largest(arr8)
    print(f"\nArray: {arr8}")
    print(f"Second largest: {result8}")
    
    # Test Case 3 - Duplicate largest
    arr9 = [10, 10, 8, 7, 5]
    result9 = find_second_largest(arr9)
    print(f"\nArray: {arr9}")
    print(f"Second largest: {result9}")
    
    # Test Case 4 - All same elements
    arr10 = [5, 5, 5, 5]
    result10 = find_second_largest(arr10)
    print(f"\nArray: {arr10}")
    print(f"Second largest: {result10}")
    
    # Test Case 5 - Two elements
    arr11 = [10, 5]
    result11 = find_second_largest(arr11)
    print(f"\nArray: {arr11}")
    print(f"Second largest: {result11}")
    
    # Test Case 6 - Negative numbers
    arr12 = [-5, -2, -10, -8, -1]
    result12 = find_second_largest(arr12)
    print(f"\nArray: {arr12}")
    print(f"Second largest: {result12}")
